### PR TITLE
refactor: implement backend based on database

### PR DESCRIPTION
**Pull request by ctron**

### CODE REVIEW

File: common/src/db.rs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+        db_name: &str,
+        db_strategy: DbStrategy,
+    ) -> Result<Self, anyhow::Error> {
+        let port = port.into().unwrap_or(5432);
+        let url = format!("postgres://{username}:{password}@{host}:{port}/{db_name}");
+        log::info!("connect to {}", url);
+
+        let mut opt = ConnectOptions::new(url);
+        opt.min_connections(16);
+        opt.sqlx_logging_level(log::LevelFilter::Trace);
```
Comment (helio-frota) [inline common/src/db.rs:95]: I remember this part :D

File: Cargo.toml [hunk 2]
```diff
@@ -6,6 +6,7 @@ members = [
   "common",
   "common/auth",
   "common/infrastructure",
+  "modules/importer",
```
Comment (helio-frota) [inline Cargo.toml:9]: :+1:

File: common/Cargo.toml [hunk 3]
```diff
@@ -3,9 +3,9 @@ name = "trustify-common"
 version = "0.1.0"
 edition = "2021"
 
-# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
-
 [dependencies]
+trustify-migration = { path = "../migration" }
```
Comment (helio-frota) [inline common/Cargo.toml:7]: :+1:

File: common/Cargo.toml [hunk 4]
```diff
@@ -14,10 +14,12 @@ log = "0.4.19"
 native-tls = "0.2"
 packageurl = "0.3.0"
 pem = "3"
+postgresql_embedded = { version = "0.6.2", features = ["blocking", "bundled", "tokio" ] }
```
Comment (helio-frota) [inline common/Cargo.toml:17]: :+1:

File: common/Cargo.toml [hunk 5]
```diff
@@ -14,10 +14,12 @@ log = "0.4.19"
...
 reqwest = { version = "0.11", features = ["native-tls"] }
 sea-orm = { version = "0.12", features = ["sea-query-binder", "sqlx-postgres", "runtime-tokio-rustls", "macros"] }
 serde = { version = "1.0.183", features = ["derive"] }
 serde_json = "1.0.104"
+tempfile = "3"
```
Comment (helio-frota) [inline common/Cargo.toml:22]: :+1:

File: common/src/db.rs [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+            ConnectionOrTransaction::Transaction(inner) => inner.query_all(stmt).await,
+        }
+    }
+}
+
+#[derive(Debug)]
+enum DbStrategy {
+    External,
+    Managed(Arc<(PostgreSQL, TempDir)>),
+}
```
Comment (helio-frota) [inline common/src/db.rs:72]: :heart_eyes_cat:

File: common/src/db.rs [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+        Migrator::refresh(&db).await?;
+        log::debug!("applied migrations");
+
+        Ok(Self {
+            db,
+            db_strategy: Arc::new(db_strategy),
+        })
+    }
+
+    pub async fn with_external_config(
```
Comment (helio-frota) [inline common/src/db.rs:109]: :heart_eyes:

File: common/src/db.rs [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+                &database.host,
+                database.port,
+                &database.name,
+                DbStrategy::External,
+            )
+            .await
+        }
+    }
+
+    pub async fn for_test(name: &str) -> Result<Self, anyhow::Error> {
```
Comment (helio-frota) [inline common/src/db.rs:136]: :heart_eyes_cat:

File: entity/Cargo.toml [hunk 9]
```diff
@@ -3,10 +3,9 @@ name = "trustify-entity"
 version = "0.1.0"
 edition = "2021"
 
-# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
-
 [dependencies]
-sea-orm = { version = "0.12", features = [ "sqlx-postgres", "runtime-tokio-rustls", "macros" ] }
+sea-orm = { version = "0.12", features = [ "sqlx-postgres", "runtime-tokio-rustls", "macros", "with-json" ] }
```
Comment (helio-frota) [inline entity/Cargo.toml:7]: :shrug:

File: graph/src/graph/advisory/advisory_vulnerability.rs [hunk 10]
```diff
@@ -1,10 +1,10 @@
-use crate::db::Transactional;
 use crate::graph::advisory::affected_package_version_range::AffectedPackageVersionRangeContext;
 use crate::graph::advisory::fixed_package_version::FixedPackageVersionContext;
 use crate::graph::advisory::not_affected_package_version::NotAffectedPackageVersionContext;
 use crate::graph::advisory::AdvisoryContext;
 use crate::graph::error::Error;
 use sea_orm::{ActiveModelTrait, ColumnTrait, EntityTrait, QueryFilter, Set};
+use trustify_common::db::Transactional;
```
Comment (helio-frota) [inline graph/src/graph/advisory/advisory_vulnerability.rs:7]: :+1:

File: graph/src/graph/advisory/csaf/mod.rs [hunk 11]
```diff
@@ -1,12 +1,12 @@
 mod util;
 
-use crate::db::Transactional;
 use crate::graph::advisory::csaf::util::resolve_purls;
 use crate::graph::advisory::AdvisoryContext;
 use crate::graph::error::Error;
 use csaf::{document::Category, Csaf};
 use sea_orm::ActiveValue::Set;
 use sea_orm::{ActiveModelTrait, TransactionTrait};
+use trustify_common::db::Transactional;
```
Comment (helio-frota) [inline graph/src/graph/advisory/csaf/mod.rs:9]: :+1:

File: graph/src/graph/mod.rs [hunk 12]
```diff
@@ -9,25 +8,19 @@ use sea_orm_migration::MigratorTrait;
...
-pub enum DbStrategy {
-    External,
-    Managed(Arc<(PostgreSQL, TempDir)>),
-}
-
 #[derive(Debug, Clone)]
 pub struct Graph {
-    db: DatabaseConnection,
-    db_strategy: Arc<DbStrategy>,
+    db: trustify_common::db::Database,
```
Comment (helio-frota) [inline graph/src/graph/mod.rs:23]: :100:

File: ingestors/Cargo.toml [hunk 13]
```diff
@@ -5,6 +5,7 @@ edition = "2021"
 publish = false
 
 [dependencies]
+trustify-common = { path = "../common"}
```
Comment (helio-frota) [inline ingestors/Cargo.toml:8]: :+1:

File: modules/importer/src/service.rs [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+            name: Set(name),
+            configuration: Set(configuration),
+        };
+
+        entity.insert(&tx).await?;
+
+        tx.commit().await?;
+
+        Ok(())
+    }
```
Comment (helio-frota) [inline modules/importer/src/service.rs:76]: clear usage +1

File: server/src/lib.rs [hunk 15]
```diff
@@ -144,18 +138,22 @@ pub fn configure(config: &mut web::ServiceConfig) {
...
-            "localhost",
-            None,
-            name,
-            DbStrategy::External,
+        db::Database::with_external_config(
+            &Database {
+                username: "postgres".to_string(),
+                password: "eggs".to_string(),
+                host: "localhost".to_string(),
+                port: 5432,
```
Comment (helio-frota) [inline server/src/lib.rs:151]: (this part of the port-config is probably related to the github-discussions env/vars etc, not a blocker-to-me :dark_sunglasses: )

**Review by helio-frota — APPROVED** [review]

File: common/src/db.rs [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+}
+
+#[derive(Clone)]
+pub enum ConnectionOrTransaction<'db> {
+    Connection(&'db DatabaseConnection),
+    Transaction(&'db DatabaseTransaction),
+}
+
+#[async_trait::async_trait]
+impl ConnectionTrait for ConnectionOrTransaction<'_> {
```
Comment (jcrossley3) [inline common/src/db.rs:31]: Why do we need to reimplement [upstream](https://docs.rs/sea-orm/latest/sea_orm/trait.ConnectionTrait.html)?

Comment (ctron) [reply common/src/db.rs:31]: That was there before, I just moved it. Not sure if we need it, I can try to remove it.

Comment (bobmcwhirter) [reply common/src/db.rs:31]: It was due to an initial foray into allowing transactional or non-transactional usage, and was probably mis-guided.

I've been back towards touching txns now, still trying to sort out better patterns.

File: common/src/db.rs [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+    }
+}
+
+#[derive(Clone)]
+pub enum ConnectionOrTransaction<'db> {
+    Connection(&'db DatabaseConnection),
+    Transaction(&'db DatabaseTransaction),
+}
+
+#[async_trait::async_trait]
```
[the remark this answers is not in the document]
Comment (jcrossley3) [reply common/src/db.rs:30]: I'm not sure who "you're" is in this case, so what is he right about?

**Pull request merged**